skbio/io/format/gff3.py

```diff
@@ -362,10 +362,8 @@ def _construct_seq(fh, constructor=DNA, seq_num=1):
     for i, (data_type, seq_id, l) in enumerate(_yield_record(fh), 1):
         if data_type == 'data' and seq_num == i:
             lines = l
-    # you can't read directly from fh because in registry.py line 543
-    # file.tell() will fail "telling position disabled by next() call".
-    stream = StringIO(fh.read())
-    seq = read(stream, format='fasta', into=constructor, seq_num=seq_num)
+    seq = _get_nth_sequence(_fasta_to_generator(fh, constructor=constructor),
```

Comment:
So technically `_fasta_to_generator` is wrapped by the decoration and will invoke the `IO` registry again. But it doesn't have any reason to sniff now which is why it is working. Regardless, I do think this is better (even though it is still re-entrant).

Looking at `fasta.py` it looks like the only way to really get the functionality you'd want without re-entering the registry would be to pull out: 
```python
for seq, id_, desc in _parse_fasta_raw(fh, _parse_sequence_data,
                                       FASTAFormatError):
    yield constructor(seq, metadata={'id': id_, 'description': desc},
                      **kwargs)
```
from `_fasta_to_generator` which probably isn't worth it, but that's up to you.

It may even be argued that decorated IO functions shouldn't actually be wrapped, which would make your code work better... I don't think there's a reason they need to be wrapped as they aren't really public functions. The registry shouldn't care either way. 

As I think about this, that is probably the way to go... So I'm good with this implementation and in the future we can just return the original function which will save some time.